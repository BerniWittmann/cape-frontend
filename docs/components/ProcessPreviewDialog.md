# ProcessPreviewDialog

A Dialog to preview a Process which shows the name, tags, description and a preview-svg. Allows to delete, edit or download a process.

## Methods

<!-- @vuese:ProcessPreviewDialog:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|closeDialog|closes the dialog|-|
|editProcess|routes to the edit page of the process|-|
|deleteProcess|deletes the process after asking for confirmation|-|

<!-- @vuese:ProcessPreviewDialog:methods:end -->


