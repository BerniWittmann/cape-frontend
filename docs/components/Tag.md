# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


