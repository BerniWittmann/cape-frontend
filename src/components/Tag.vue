<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
