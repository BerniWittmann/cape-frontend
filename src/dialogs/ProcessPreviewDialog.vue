<template>
  <el-dialog :title="activeProcess.name" :visible.sync="showProcessPreview">
    <el-row>
      <el-col>
        <tag v-for="tag in activeProcess.tags" :tag="tag" :key="tag.id" size="small"></tag>
      </el-col>
    </el-row>
    <el-row class="margin-top">
      <el-col>
        <span>{{activeProcess.description}}</span>
      </el-col>
    </el-row>
    <br>
    <iframe v-bind:style="{ height: svgHeightAdjust }" :srcdoc="processSVG" class="svg-view margin-top"></iframe>
    <el-row class="margin-top">
      <el-col>
        <el-button icon="el-icon-edit" size="small" @click="editProcess">{{ $t('process.edit.link') }}
        </el-button>
        <el-button icon="el-icon-delete" type="danger" plain size="small" @click="deleteProcess">{{
          $t('process.delete.delete_button') }}
        </el-button>
        <a :href="fileContent" :download="fileName" class="download-link"><el-button plain size="small" icon="el-icon-download">{{$t('process.edit.download')}}</el-button></a>
      </el-col>
    </el-row>
  </el-dialog>
</template>

<script>
/* ============
    * Process Preview Dialog
    * ============
    *
    * Shows a preview of the Dialog
    */
import Tag from '@/components/Tag.vue'
import processService from '@/services/process'

export default {
  components: {
    Tag
  },

  computed: {
    activeProcess() {
      return this.$store.state.process.activeProcess
    },
    showProcessPreview: {
      get: function () {
        return !!this.activeProcess
      },
      set: function () {
        this.closeDialog()
      }
    },
    svgHeightAdjust() {
      if (!this.activeProcess.svg) return 'auto'
      const maxHeight = 500
      const heightPos = this.activeProcess.svg.indexOf('height="') + 8
      const height = parseInt(this.activeProcess.svg.substring(heightPos, this.activeProcess.svg.indexOf('"', heightPos)))
      return height <= maxHeight ? height + 20 + 'px' : maxHeight + 'px'
    },
    fileContent() {
      return 'data:text/xml,' + encodeURIComponent(this.activeProcess.xml)
    },
    fileName() {
      return this.activeProcess.name.replace(/ /g, '_') + '.bpmn'
    },
    processSVG() {
      return !this.activeProcess.svg ? this.$t('process.error_svg_preview') : this.activeProcess.svg
    }
  },

  methods: {
    closeDialog() {
      this.$router.back()
    },
    editProcess() {
      this.$router.push({
        name: 'process.edit',
        params: this.$route.params
      })
    },

    deleteProcess() {
      this.$confirm(this.$t('process.delete.message'), this.$t('process.delete.warning'), {
        confirmButtonText: this.$t('process.delete.ok'),
        cancelButtonText: this.$t('process.delete.cancel'),
        type: 'warning',
        cancelButtonClass: 'is-plain el-button--info',
        confirmButtonClass: 'el-button--danger'
      }).then(() => {
        this.$message({
          type: 'success',
          message: this.$t('process.delete.confirmation')
        })
        processService.remove(this.activeProcess)
        this.$router.back()
      }).catch(() => {
        this.$message({
          type: 'info',
          message: this.$t('process.delete.cancellation')
        })
      })
    }
  }

}
</script>

<style scoped lang="scss">

.svg-view {
  width: 100%;
  height: 100%;
  min-height: 100%;
  border: none;
}

.margin-top {
  margin-top: 15px;
}

.download-link {
  margin-left: 10px;
}

</style>
