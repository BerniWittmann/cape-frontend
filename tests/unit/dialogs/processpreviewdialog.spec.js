import { mount, shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import moment from 'moment'
import { Button, Dialog } from 'element-ui'

import Tag from '@/components/Tag.vue'
import ProcessPreviewDialog from '@/dialogs/ProcessPreviewDialog.vue'
import ProcessService from '@/services/process'

describe('Dialogs', () => {
  describe('Process Preview Dialog', () => {
    let store
    let cmp

    const date = moment(moment().utc())
    const router = {
      back: jest.fn(),
      push: jest.fn()
    }
    const route = {
      params: {
        processID: 42
      }
    }
    beforeEach(() => {
      store = {
        state: {
          process: {
            activeProcess: {
              id: '1',
              name: 'My Process',
              createdAt: date.clone().subtract(14, 'days'),
              lastEditedAt: date.clone().subtract(2, 'days'),
              svg: 'text height="30" more text="even more Text"',
              description: 'My Description',
              xml: '<?xml version="1.0" encoding="UTF-8"?><bpmn:Process>Test Process</bpmn:Process>',
              tags: [{
                id: '42',
                name: 'First Tag',
                color: '#FF0000'
              }, {
                id: '43',
                name: 'Second Tag',
                color: '#FFFF00'
              }]
            }
          }
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(ProcessPreviewDialog, {
        i18n,
        mocks: {
          $store: store,
          $router: router,
          $route: route
        },
        stubs: {
          ElButton: Button,
          ElDialog: Dialog
        }
      })
    }

    it('can render', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('can render tags', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)
      expect(tags.at(0).props('tag')).toEqual(store.state.process.activeProcess.tags[0])
      expect(tags.at(1).props('tag')).toEqual(store.state.process.activeProcess.tags[1])
    })

    it('can show the name', () => {
      expect(cmp.find('.el-dialog__title').text()).toEqual(store.state.process.activeProcess.name)
    })

    it('can show the svg in iframe', () => {
      expect(cmp.find('iframe').attributes('srcdoc')).toEqual(store.state.process.activeProcess.svg)
    })

    it('can extract and reuse the height from svg', () => {
      expect(cmp.find('iframe').attributes('style')).toEqual('height: 50px;')
      store = {
        state: {
          process: {
            activeProcess: {
              id: '42',
              name: 'My Process 1',
              createdAt: date.clone().subtract(14, 'days'),
              lastEditedAt: date.clone().subtract(2, 'days'),
              svg: 'placeholder height="3000" placeholder',
              tags: [{
                id: '421',
                name: 'First Tag',
                color: '#FF0000'
              }, {
                id: '43',
                name: 'Second Tag',
                color: '#FFFF00'
              }]
            }
          }
        }
      }
      cmp = mount(ProcessPreviewDialog, {
        i18n,
        mocks: {
          $store: store
        }
      })
      expect(cmp.vm.svgHeightAdjust).toBe('500px')
    })

    it('renders the description', () => {
      const description = cmp.find('elcol-stub span')
      expect(description.text()).toEqual(store.state.process.activeProcess.description)
    })
    it('can be closed', () => {
      expect(cmp.html()).toMatchSnapshot()
      cmp.find('.el-dialog__header button').trigger('click')
      expect(router.back).toHaveBeenCalled()
    })

    it('can be edited', () => {
      expect(cmp.html()).toMatchSnapshot()
      const button = cmp.find('elrow-stub button')
      expect(button.exists()).toBeTruthy()
      button.trigger('click')
      expect(router.push).toHaveBeenCalledWith({
        name: 'process.edit',
        params: route.params
      })
    })

    describe('can be deleted', () => {
      beforeEach(() => {
        cmp.vm.$confirm = jest.fn().mockImplementation(() => ({
          then: (arg) => {
            return {
              catch: () => {
              }
            }
          }
        }))
        ProcessService.remove = jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
        cmp.vm.$message = jest.fn()
      })
      it('show as confirmation dialog', () => {
        expect(cmp.html()).toMatchSnapshot()
        const button = cmp.findAll('elrow-stub button').at(1)
        expect(button.exists()).toBeTruthy()
        button.trigger('click')
        expect(cmp.vm.$confirm).toHaveBeenCalledWith('process.delete.message', 'process.delete.warning',
          {
            'cancelButtonText': 'process.delete.cancel',
            'confirmButtonText': 'process.delete.ok',
            'type': 'warning',
            'cancelButtonClass': 'is-plain el-button--info',
            'confirmButtonClass': 'el-button--danger'
          })
      })
      it('deletes the process on confirm  ', () => {
        cmp.vm.$confirm = jest.fn().mockImplementation(() => ({
          then: (arg) => {
            arg()
            return {
              catch: () => {
              }
            }
          }
        }))
        cmp.findAll('elrow-stub button').at(1).trigger('click')
        expect(ProcessService.remove).toHaveBeenCalledWith(store.state.process.activeProcess)
        expect(cmp.vm.$message).toHaveBeenCalledWith({ 'message': 'process.delete.confirmation', 'type': 'success' })
        expect(router.back).toHaveBeenCalled()
      })
      it('deletes the process on confirm  ', () => {
        cmp.vm.$confirm = jest.fn().mockImplementation(() => ({
          then: () => {
            return {
              catch: (arg) => {
                arg()
              }
            }
          }
        }))
        cmp.findAll('elrow-stub button').at(1).trigger('click')
        expect(cmp.vm.$message).toHaveBeenCalledWith({ 'message': 'process.delete.cancellation', 'type': 'info' })
      })
    })
  })
})
