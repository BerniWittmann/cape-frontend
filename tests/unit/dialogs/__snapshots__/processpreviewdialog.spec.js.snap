// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogs Process Preview Dialog can be closed 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2027;">
  <div role="dialog" aria-modal="true" aria-label="My Process" class="el-dialog" style="margin-top: 15vh;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Process</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elrow-stub tag="div" justify="start" align="top">
        <elcol-stub span="24" tag="div">
          <tag-stub tag="[object Object]" size="small"></tag-stub>
          <tag-stub tag="[object Object]" size="small"></tag-stub>
        </elcol-stub>
      </elrow-stub>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><span>My Description</span></elcol-stub>
      </elrow-stub> <br> <iframe srcdoc="text height=&quot;30&quot; more text=&quot;even more Text&quot;" class="svg-view margin-top" style="height: 50px;"></iframe>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><button type="button" class="el-button el-button--default el-button--small">
            <!----><i class="el-icon-edit"></i><span>process.edit.link
      </span></button> <button type="button" class="el-button el-button--danger el-button--small is-plain">
            <!----><i class="el-icon-delete"></i><span>process.delete.delete_button
      </span></button> <a href="data:text/xml,%3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22UTF-8%22%3F%3E%3Cbpmn%3AProcess%3ETest%20Process%3C%2Fbpmn%3AProcess%3E" download="My_Process.bpmn" class="download-link"><button type="button" class="el-button el-button--default el-button--small is-plain">
              <!----><i class="el-icon-download"></i><span>process.edit.download</span></button></a></elcol-stub>
      </elrow-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Dialogs Process Preview Dialog can be deleted show as confirmation dialog 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2031;">
  <div role="dialog" aria-modal="true" aria-label="My Process" class="el-dialog" style="margin-top: 15vh;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Process</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elrow-stub tag="div" justify="start" align="top">
        <elcol-stub span="24" tag="div">
          <tag-stub tag="[object Object]" size="small"></tag-stub>
          <tag-stub tag="[object Object]" size="small"></tag-stub>
        </elcol-stub>
      </elrow-stub>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><span>My Description</span></elcol-stub>
      </elrow-stub> <br> <iframe srcdoc="text height=&quot;30&quot; more text=&quot;even more Text&quot;" class="svg-view margin-top" style="height: 50px;"></iframe>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><button type="button" class="el-button el-button--default el-button--small">
            <!----><i class="el-icon-edit"></i><span>process.edit.link
      </span></button> <button type="button" class="el-button el-button--danger el-button--small is-plain">
            <!----><i class="el-icon-delete"></i><span>process.delete.delete_button
      </span></button> <a href="data:text/xml,%3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22UTF-8%22%3F%3E%3Cbpmn%3AProcess%3ETest%20Process%3C%2Fbpmn%3AProcess%3E" download="My_Process.bpmn" class="download-link"><button type="button" class="el-button el-button--default el-button--small is-plain">
              <!----><i class="el-icon-download"></i><span>process.edit.download</span></button></a></elcol-stub>
      </elrow-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Dialogs Process Preview Dialog can be edited 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2029;">
  <div role="dialog" aria-modal="true" aria-label="My Process" class="el-dialog" style="margin-top: 15vh;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Process</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elrow-stub tag="div" justify="start" align="top">
        <elcol-stub span="24" tag="div">
          <tag-stub tag="[object Object]" size="small"></tag-stub>
          <tag-stub tag="[object Object]" size="small"></tag-stub>
        </elcol-stub>
      </elrow-stub>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><span>My Description</span></elcol-stub>
      </elrow-stub> <br> <iframe srcdoc="text height=&quot;30&quot; more text=&quot;even more Text&quot;" class="svg-view margin-top" style="height: 50px;"></iframe>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><button type="button" class="el-button el-button--default el-button--small">
            <!----><i class="el-icon-edit"></i><span>process.edit.link
      </span></button> <button type="button" class="el-button el-button--danger el-button--small is-plain">
            <!----><i class="el-icon-delete"></i><span>process.delete.delete_button
      </span></button> <a href="data:text/xml,%3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22UTF-8%22%3F%3E%3Cbpmn%3AProcess%3ETest%20Process%3C%2Fbpmn%3AProcess%3E" download="My_Process.bpmn" class="download-link"><button type="button" class="el-button el-button--default el-button--small is-plain">
              <!----><i class="el-icon-download"></i><span>process.edit.download</span></button></a></elcol-stub>
      </elrow-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Dialogs Process Preview Dialog can render 1`] = `
<div class="el-dialog__wrapper" name="dialog-fade" style="z-index: 2001;">
  <div role="dialog" aria-modal="true" aria-label="My Process" class="el-dialog" style="margin-top: 15vh;">
    <div class="el-dialog__header"><span class="el-dialog__title">My Process</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
    <div class="el-dialog__body">
      <elrow-stub tag="div" justify="start" align="top">
        <elcol-stub span="24" tag="div">
          <tag-stub tag="[object Object]" size="small"></tag-stub>
          <tag-stub tag="[object Object]" size="small"></tag-stub>
        </elcol-stub>
      </elrow-stub>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><span>My Description</span></elcol-stub>
      </elrow-stub> <br> <iframe srcdoc="text height=&quot;30&quot; more text=&quot;even more Text&quot;" class="svg-view margin-top" style="height: 50px;"></iframe>
      <elrow-stub tag="div" justify="start" align="top" class="margin-top">
        <elcol-stub span="24" tag="div"><button type="button" class="el-button el-button--default el-button--small">
            <!----><i class="el-icon-edit"></i><span>process.edit.link
      </span></button> <button type="button" class="el-button el-button--danger el-button--small is-plain">
            <!----><i class="el-icon-delete"></i><span>process.delete.delete_button
      </span></button> <a href="data:text/xml,%3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22UTF-8%22%3F%3E%3Cbpmn%3AProcess%3ETest%20Process%3C%2Fbpmn%3AProcess%3E" download="My_Process.bpmn" class="download-link"><button type="button" class="el-button el-button--default el-button--small is-plain">
              <!----><i class="el-icon-download"></i><span>process.edit.download</span></button></a></elcol-stub>
      </elrow-stub>
    </div>
    <!---->
  </div>
</div>
`;
